src/ergw_tei_mngr.erl

```diff
@@ -0,0 +1,174 @@
+
+    Data = #data{node_id = NodeId,
+		  block_id = BlockId,
+		  prefix = Prefix,
+		  prefix_len = PrefixLen,
+		  tid = TID,
+		  save_cnt = ?SAVE_COUNTER},
+
+    {ok, #{}, Data}.
+
+handle_event(state_timeout, save, State, Data) ->
+    save_data(State, Data),
+    {keep_state, Data#data{save_cnt = ?SAVE_COUNTER}};
+
+handle_event({call, From}, {alloc_tei, Name, KeyFun}, State, Data) ->
+    RndState = maybe_init_rnd(Name, State),
+    alloc_tei(From, Name, KeyFun, ?MAX_TRIES, RndState, State, Data);
+
+handle_event({call, From}, {release_tei, _Name, TEI, KeyFun}, _State, #data{tid = TID}) ->
+    ets:delete(TID, KeyFun(TEI)),
+    Actions = [{reply, From, ok}, state_timeout()],
+    {keep_state_and_data, Actions};
+
+handle_event({call, From}, Request, _State, _Data) ->
+    Reply = {error, {badarg, Request}},
+    {keep_state_and_data, [{reply, From, Reply}]};
+
+handle_event(_, _Msg, _State, _Data) ->
+    keep_state_and_data.
+
+terminate(_Reason, _State, _Data) ->
+    ok.
+
+code_change(_OldVsn, State, Data, _Extra) ->
+    {ok, State, Data}.
+
+%%%=========================================================================
+%%%  internal functions
+%%%=========================================================================
+
+maybe_init_rnd(Name, State) when is_map_key(Name, State) ->
+    maps:get(Name, State);
+maybe_init_rnd(_, _) ->
+    rand:seed_s(exrop).
+
+state_timeout() ->
+    {state_timeout, ?SAVE_TIMEOUT, save}.
+
+serialize_data(State, #data{tid = TID}) ->
+    SeedL = maps:fold(fun(K,V,L) -> [{K, rand:export_seed_s(V)}|L] end, [], State),
+    Term =
+	#{seed  => SeedL,
+	  teids => ets:tab2list(TID)},
+    term_to_binary(Term, [{minor_version, 2}, compressed]).
+
+save_data(State, #data{node_id = _NodeId, block_id = _BlockId} = Data) ->
+    _Serialized = serialize_data(State, Data).
+
+alloc_tei_reply(From, Reply, Name, RndState, State, Data) ->
+    Actions = [{reply, From, Reply}, state_timeout()],
+    {next_state, State#{Name => RndState}, Data, Actions}.
+
+alloc_tei(From, Name, _KeyFun, 0, RndState, State, Data) ->
+    alloc_tei_reply(From, {error, no_tei}, Name, RndState, State, Data);
+alloc_tei(From, Name, KeyFun, Cnt, RndState0, State,
+	  #data{prefix = Prefix0, prefix_len = PrefixLen, tid = TID} = Data) ->
+    Prefix = Prefix0 bsl (?MAX_TRIES - PrefixLen),
+    Mask = 16#ffffffff bsr PrefixLen,
+    {TEI0, RndState} = rand:uniform_s(16#fffffffe, RndState0),
+    TEI = Prefix + (TEI0 band Mask),
+    case ets:insert_new(TID, {KeyFun(TEI), Name}) of
+	true ->
+	    alloc_tei_reply(From, {ok, TEI}, Name, RndState, State, Data);
+	false ->
+	    alloc_tei(From, Name, KeyFun, Cnt - 1, RndState, State, Data)
+    end.
+
+get_teid_config() ->
+    case application:get_env(ergw, teid) of
+        {ok, {_, _} = Config} ->
+            Config;
+        {ok, {Len}} ->
+            {?POW(2, Len), Len};
+        _ ->
+            {1, 0}
+    end.
```

Comment:
validation phase should take care of that. The init method should be able to assume that the config key always contains sensible data without this case here.

Reply:
We have a 3 cases:
1 - when set Length and Prefix
2 - when set Length only
3 - when configuration is now a set at all
Here we don't do validation, we try to create default value for Prefix set default Length and Prefix

Reply:
your implementation logic always expect a tuple `{Length, Prefix}`. And that is all it should care about. The logic of having a default or just a single number, should go into the validation step.